pkg/storage/driver/multisecrets.go

```diff
@@ -0,0 +1,438 @@
+	// delete remaining chunks if any
+	for chunk := 2; chunk <= chunks; chunk++ {
+		err = secrets.impl.Delete(context.Background(), fmt.Sprintf("%s.%d", key, chunk), metav1.DeleteOptions{})
+	}
+	return rls, err
+}
+
+// newMultiSecretsObject constructs a kubernetes Secret object
+// to store a release. Each secret data entry is the base64
+// encoded gzipped string of a release.
+//
+// The following labels are used within each secret:
+//
+//	"modifiedAt"    - timestamp indicating when this secret was last modified. (set in Update)
+//	"createdAt"     - timestamp indicating when this secret was created. (set in Create)
+//	"version"        - version of the release.
+//	"status"         - status of the release (see pkg/release/status.go for variants)
+//	"owner"          - owner of the secret, currently "helm".
+//	"name"           - name of the release.
+func newMultiSecretsObject(key string, rls *rspb.Release, lbs labels, chunkSizeExist int) (*[]v1.Secret, error) {
+	const owner = "helm"
+
+	// encode the release
+	s, err := encodeRelease(rls)
+	if err != nil {
+		return nil, err
+	}
+
+	if lbs == nil {
+		lbs.init()
+	}
+
+	// apply labels
+	lbs.set("name", rls.Name)
+	lbs.set("owner", owner)
+	lbs.set("status", rls.Info.Status.String())
+	lbs.set("version", strconv.Itoa(rls.Version))
+
+	// create and return secret object.
+	// Helm 3 introduced setting the 'Type' field
+	// in the Kubernetes storage object.
+	// Helm defines the field content as follows:
+	// <helm_domain>/<helm_object>.v<helm_object_version>
+	// Type field for Helm 3: helm.sh/release.v1
+	// Note: Version starts at 'v1' for Helm 3 and
+	// should be incremented if the release object
+	// metadata is modified.
+	// This would potentially be a breaking change
+	// and should only happen between major versions.
+
+	chunkSize, _ := getMaxChunkSize(key, nil, chunkSizeExist)
+
+	objs := []v1.Secret{}
+	// Needs chunking?
+	if len(s) > chunkSize {
+		origData := s
+
+		lbs.set("chunksize", strconv.Itoa(chunkSize))
+
+		slices := []string{}
+		lastIndex := 0
+		lastI := 0
+		for i := range origData {
+			if i-lastIndex > chunkSize {
+				slices = append(slices, origData[lastIndex:lastI])
+				lastIndex = lastI
+			}
+			lastI = i
+		}
+		// handle the leftover data at the end
+		if len(origData)-lastIndex > chunkSize {
+			slices = append(slices, origData[lastIndex:lastIndex+chunkSize], origData[lastIndex+chunkSize:])
+		} else {
+			slices = append(slices, origData[lastIndex:])
+		}
+
+		i := 1
+		for _, str := range slices {
+			{
+				var lbs2 labels
+				lbs2.init()
+				lbs2.set("chunk", strconv.Itoa(i))
+				lbs2.set("chunks", strconv.Itoa(len(slices)))
+				instanceName := key
+				if i > 1 {
+					instanceName = fmt.Sprintf("%s.%d", key, i)
+				}
+				objs = append(objs, v1.Secret{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:   instanceName,
+						Labels: kblabels.Merge(lbs.toMap(), lbs2.toMap()),
+					},
+					Type: "helm.sh/release.v1",
```

Comment:
@mattfarina with current implementation,  if size <= 10M, "hem list" with HELM_DRIVER="secret" contains all of releases  created by "secret" and "multisecret". if we introduce new type "helm.sh/release.chunk.v1", for single release whose size >1MB, first secret data will be labelled by "helm.sh/release.v", others will be labelled by "helm.sh/release.chunk.v1", it is a little bit strange. besides, we dont' benefit from new secret type.  

currently,  with HELM_DRIVER="multisecret" , "helm list" will contains releases created by "secret" and "multisecret"
with HELM_DRIVER="secret", it will list the releases created with "secret", "multisecret" whose size is <1MB.  if we want to  support the releases created with "mutlisecret",  we need change existing helm secret driver logic.  but if that is case, why not make "multisecret" as default.  currently, i think with "HLEM_DRIVER=secret", it needn't take care of the secretd created with "multisecret" since it is not supported by itself.  what do you think?

Reply:
Indeed critical to consider the behavior of old and new clients in parallel, @mattfarina!   

The current design considers the situation that helm charts can grow over a period of time, and **only eventually run into the size limit**.   Once this point is reached with a new chart version, the use of the chunked-secret storage is mandatory for all users who manage that release (chart/values combination), a user with an old client would not be able to manage this application any more in any case.   

I'd assume that if the multisecret driver introduced a completely new type, then either:
-  We could constrain that a switch from secret to multisecret is not even possible - this would kill the upgrade experience and force users to completely restart with a fresh install of release-1, as opposed to smoothly transition from their existing single-secret release version history to now include a multi-secret release version (and potentially back, if the size reduced later on again in a newer release)
- If we allowed the transition from single-secret to multi-secret storage but eclipsed those release versions that stored in chunks, then a user with an old client would see an incomplete version history, and potentially misinterpret what the deployed version actually is (including retrieving wrong values or basing upgrades on wrong version)

I believe both of these approaches would be worse for users than requiring that two users who jointly manage an instance of a deployed application to agree on using the new helm version with the chunked storage feature.  Ultimately, the app team that provides such large helm charts would have to give respective instructions to their users, i.e. I cannot imagine that one user knows of the need for multi-secret, and  the other doesn't.

I wonder if we could place something in the release info (e.g. in the release description) that would indicate the switch to multisecret mode, such that a helm history with an old client could still display the (multisecret) release as a line item with a human-readable text ("upgrade completed to multisecret storage") that would complement the potential error that an old client would show it someone tried to do a values retrieval or upgrade based on that version?

Reply:
@mattfarina  as an update to your question on coexistence between multisecret-created helm releases and old helm clients that don't have multisecret support, I have discussed an approach with @joejulian where the multisecret driver uses regular helm.sh/release.v1-type secrets if chunking is not required (the majority of cases), and uses a new secret-type helm.sh/release-multisecret.v1 in those cases where chunking is required because of the 1-MB limit.  The following is the experience for users of an old helm client if a release uses the new type:

Helm list does not show the release if the "deployed" release is split across multiple secrets and thus cannot be parsed by the secret storage driver.  No cryptic error message, good:  

# helm list -n apptest
NAME    NAMESPACE       REVISION        UPDATED STATUS  CHART   APP VERSION


If the user wonders about where the release is gone, inspecting the release storage gives a good indication with the different type:

# kubectl get secret -n apptest | grep sh.helm
sh.helm.release.v1.apptest.v1                       helm.sh/release.v1           1      46h
sh.helm.release.v1.apptest.v2                       helm.sh/release.v1           1      46h
sh.helm.release.v1.apptest.v3                       helm.sh/release.v1           1      46h
sh.helm.release.v1.apptest.v4                       helm.sh/release.v1           1      46h
sh.helm.release.v1.apptest.v5                       helm.sh/release.v1           1      46h
sh.helm.release.v1.apptest.v6                       helm.sh/release-multisecret.v1   1      25h
sh.helm.release.v1.apptest.v6.2                     helm.sh/release-multisecret.v1   1      25h


Helm history works to the level that it lists all non-chunked releases and shows the correct status "superseded".  Again, if a user wonders why the "deployed" release does not show, an inspection of the Kube secrets would show:

# helm history apptest -n apptest
REVISION        UPDATED                         STATUS          CHART                   APP VERSION     DESCRIPTION
1               Mon Sep  9 16:56:11 2024        superseded      apptest-0.0.1      0.0.1           Install complete
2               Mon Sep  9 17:03:24 2024        superseded      apptest-0.0.1      0.0.1           Upgrade complete
3               Mon Sep  9 17:06:40 2024        superseded      apptest-0.0.1      0.0.1           Upgrade complete
4               Mon Sep  9 17:13:27 2024        superseded      apptest-0.0.1      0.0.1           Upgrade complete
5               Mon Sep  9 17:17:04 2024        superseded      apptest-0.0.1      0.0.1           Upgrade complete


@joejulian had the concern if the user could accidentally install an application with the same name into the namespace.  That concern has not materialized, the old client shows the ergular error message that comes up in these cases, good:

# helm install apptest  -n apptest apptest -f my-values2.yaml
Error: INSTALLATION FAILED: cannot re-use a name that is still in use